contracts/Marketplace2.sol

```diff
@@ -0,0 +1,420 @@
+
+import "./PurchaseListener.sol";
+import "./Ownable.sol";
+
+
+contract IMarketplace {
+    enum ProductState {
+        NotDeployed,                // non-existent or deleted
+        Deployed                    // created or redeployed
+    }
+
+    enum Currency {
+        DATA,                       // "token wei" (10^-18 DATA)
+        USD                         // attodollars (10^-18 USD)
+    }
+
+    function getProduct(bytes32 id) public view returns (string name, address owner, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds, ProductState state) {}
+    function getSubscription(bytes32 productId, address subscriber) public view returns (bool isValid, uint endTimestamp) {}
+}
+
+/**
+ * @title Streamr Marketplace
+ * @dev note about numbers:
+ *   All prices and exchange rates are in "decimal fixed-point", that is, scaled by 10^18, like ETH vs wei.
+ *  Seconds are integers as usual.
+ *
+ * Next version TODO:
+ *  1. Bancor integration
+ *  - EIP-165 inferface definition; PurchaseListener
+ *  - struct creation should use named arguments: Product({id: id, ... })
+ *  - buyFor: buy could specify who gets to subscription (without transferSubscription)
+ *  - grantSubscription: product owner could give a subscription to X
+ */
+contract Marketplace2 is Ownable, IMarketplace {
+    using SafeMath for uint256;
+
+    // product events
+    event ProductCreated(address indexed owner, bytes32 indexed id, string name, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds);
+    event ProductUpdated(address indexed owner, bytes32 indexed id, string name, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds);
+    event ProductDeleted(address indexed owner, bytes32 indexed id, string name, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds);
+    event ProductImported(address indexed owner, bytes32 indexed id, string name, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds);
+    event ProductRedeployed(address indexed owner, bytes32 indexed id, string name, address beneficiary, uint pricePerSecond, Currency currency, uint minimumSubscriptionSeconds);
+    event ProductOwnershipOffered(address indexed owner, bytes32 indexed id, address indexed to);
+    event ProductOwnershipChanged(address indexed newOwner, bytes32 indexed id, address indexed oldOwner);
+
+    // subscription events
+    event Subscribed(bytes32 indexed productId, address indexed subscriber, uint endTimestamp);
+    event NewSubscription(bytes32 indexed productId, address indexed subscriber, uint endTimestamp);
+    event SubscriptionExtended(bytes32 indexed productId, address indexed subscriber, uint endTimestamp);
+    event SubscriptionImported(bytes32 indexed productId, address indexed subscriber, uint endTimestamp);
```

Comment:
do we want to emit a SubscriptionImported event when subscription is imported?

Reply:
Yes, I don't see why not